oscillator/solver-python/oscillator.py

```diff
@@ -133,22 +133,22 @@ class Participant(Enum):
     precice_dt = participant.get_max_time_step_size()
     dt = np.min([precice_dt, my_dt])
 
-    read_times = time_stepper.rhs_eval_points(dt)
-    f = len(read_times) * [None]
-
-    for i in range(len(read_times)):
-        read_data = participant.read_data(mesh_name, read_data_name, vertex_ids, read_times[i])
-        f[i] = read_data[0]
+    def f(t): return participant.read_data(mesh_name, read_data_name, vertex_ids, t)[0]
 
     # do generalized alpha step
-    u_new, v_new, a_new = time_stepper.do_step(u, v, a, f, dt)
+    u_new, v_new, a_new, sol = time_stepper.do_step(u, v, a, f, dt)
     t_new = t + dt
 
-    write_data = [connecting_spring.k * u_new]
-
-    participant.write_data(mesh_name, write_data_name, vertex_ids, write_data)
-
-    participant.advance(dt)
+    n_pseudo = 5 * len(sol.ts)  # 5 times no. substeps should be on the safe side.
```

Comment:
Could you turn the "5" into a variable and iterate on the comment. I guess this is the most important part to understand the magic here.

Reply:
```suggestion
    samples_per_step = 5  # create n samples_per_step of time stepping scheme. Degree of dense interpolating function is usually larger 1 and, therefore, we need multiple samples per step.
    n_time_steps = len(sol.ts)  # number of time steps performed by adaptive time stepper
    n_pseudo = samples_per_step * n_time_steps  # samples_per_step times no. time steps per window.
```
Is this clearer? Ideally we would not do a uniform sampling (`dt_pseudo = dt / n_pseudo`) below but sample non-uniformly depending on `sol.ts`. But this is future work :smirk:

Reply:
yes, good now :+1: